sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3878,7 +3878,7 @@ class Dataset[T] private[sql](
    */
   def persist(newLevel: StorageLevel): this.type = {
     sparkSession.sharedState.cacheManager.cacheQuery(this, None, newLevel)
-    this
+    Dataset(sparkSession, this.queryExecution.logical).asInstanceOf[this.type]
```

Comment:
BTW, the original code came from #4686 (at least) and seems to be the default Apache Spark behavior for a long time.

Reply:
Hmm...Sorry I do not understand your concerns. It does not change the immutability of cached data.